### PR TITLE
raftstore: Implement coprocessor observer on_empty_cmd

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What is changed and how it works?

Issue Number: ref #12849

What's Changed:

```commit-message
Support new observers on_empty_cmd.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by BusyJay** [comment]
Any test case?

**Comment by CalvinNeo** [comment]
> Any test case?

I have tests in `test_trigger_right_hook`. Since `on_empty_cmd` will not affect any procedures, so I think we only need to ensure it is called at the right time.

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -1077,7 +1077,9 @@ where
 
             return self.process_raft_cmd(apply_ctx, index, term, cmd);
         }
-        // TOOD(cdc): should we observe empty cmd, aka leader change?
+
+        // we should observe empty cmd, aka leader change
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:1081]: Inaccurate comment. Empty command doesn't have to be associated with leadership change.

Comment (CalvinNeo) [reply components/raftstore/src/store/fsm/apply.rs]: try to fix

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Review by BusyJay — APPROVED** [review]

**Pull request merged**